### ISSUE TITLE
Naming of Spack commands

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#4159]
I had an interesting conversation the other day with a professor from MIT who is researching software design. He did a case study on git and found that the most reliable way of determining whether or not a command is well named is by asking multiple users exactly what they think it does. If everyone gives a different answer, it isn't well designed.

I'm opening this issue for a candid discussion on whether or not our current Spack commands are aptly named and whether or not some of them should be merged. This will probably lead to many flame wars, so please treat each suggestion with constructive criticism. Even if you don't use some of these commands (I certainly don't), there are many users who do rely on them.

### `list` vs `find`

Okay, this is by far the most controversial change I'm suggesting. But I just want to make it clear that people who are switching to Spack from other package managers will routinely mix up these commands.

In this case, every package manager uses a different command. For Homebrew, `brew list` lists installed packages, while `brew search` lists available packages. For `dnf/yum`, `dnf list` seems to be used for listing available _and_ installed packages, while `dnf search` seems to be used for searching for specific packages. For Modules, `module list` lists loaded modules, while `module avail` lists available modules. 

I'm not sure if there are better names we could use, or if they are fine as is. So I'm open to suggestions on this one.

###  `module load` vs `spack load` vs `spack module load`

Disclaimer: I don't use these commands

As someone who doesn't use these commands, I don't really know what the difference is between these three things. Personally, I plan on sticking with `module load`. I'm fine with having a special Spack command that accepts specs, but we should only have one.

### `clean` vs `purge`

I mentioned this in #2942, but we should probably merge these commands. Thoughts on whether to call the command `clean` or `purge`? Or should `spack clean` with no argument clean all stages and `spack purge` be used to purge things like caches and databases? I actually like the latter idea more now that I think about it.

EDIT: These commands were merged in #4970.

### `config` vs `configure`

Here we have two completely unrelated commands with similar names. It might be good to get rid of `config` and use `spack compiler edit` or `spack external get` do the same job (we are talking about adding a `spack external` command to add external packages in #2507).
 
### `diy` vs `setup`
(did there used to be a `spconfig` command too?)

Disclaimer: I don't use these commands

These commands seem incredibly useful for developers, but based on the name alone I have no idea what the difference is. I was talking to Jim Amundson (@amundson?) the other day about a different developer command that he was working on (although not planning on contributing to Spack). I think @citibeth mentioned that there was some Spack Environments working group tasked with solving this problem, so I'll let them handle this.

### `test` vs `install --run-tests`

I think users who want to test their installations will be confused that `spack test` doesn't do what they think it does. This is probably an issue for a later date since we have a lot of current work on this front and things will likely change.

### `md5` vs `sha256` vs `checksum`

I think @citibeth has mentioned this before, but we should probably merge these commands.

EDIT: #6428 removed the `md5` and `sha256` commands.

### `doc` vs `docs`

#3033 (currently) adds a `spack docs` command. There is already an existing `spack doc` command. As you can imagine, this is pretty confusing.

EDIT: `doc` was renamed `pydoc`.

### aliases

Do we really need `spack compiler list` _and_ `spack compilers`? Also, why are there so many aliases for the subcommands of `spack view`? Every subcommand has 2 or 3 aliases:
```
$ spack view --help
usage: spack view [-h] [-v] [-e EXCLUDE] [-d {true,false,yes,no}] ACTION ...

positional arguments:
  ACTION
    symlink (add, soft)
                        add package files to a filesystem view via symbolic
                        links
    hardlink (hard)     add packages files to a filesystem via via hard links
    remove (rm)         remove packages from a filesystem view
    statlink (status, check)
                        check status of packages in a filesystem view
```

**Comment by tgamblin** [comment]
* `list` vs `find`: I'm at a loss too.  I think `avail` and `list` might be the most clear, but I don't think anyone except someone who already uses modules thinks of `avail`.  I think most people want to type `list`.  As long as we're being nitpicky, I think `search` is slower to type for most people because it's more letters, and the `searc` part is kind of a crampy sequence that has to be done on your left hand (Dvorak users likely have different opinions).

* `module` load vs `spack load` vs `spack module loads`: it's even more confusing because it's `module loads`.  Personally I think we should do a better job of generating modules and make `module load` do prereqs by default.  I think `spack load` will be less important once we have environments.

* `clean` vs `purge`: agree. PR welcome.

* `test` vs. `install --run-tests`: `test` should become a user-facing command to run package tests, and the current functionality could be moved to `spack unit-test` or something.

* `md5` vs. `checksum`: These don't really do the same thing.  I think `spack md5` should go away, `spack checksum` should do what it does with a sensible default hash (SHA256 or SHA512) and `spack checksum` should be changed to `spack find-new-versions`  or something more intuitive.  It could be an option on `spack checksum`, I suppose, but I feel like it's different enough in that it actually spiders the web that it should have its own command.

* `doc` vs. `docs`: changing to `pydoc` and `docs`.

* `aliases`: I don't really have an opinion on this, except that I like typing `spack compilers` much better than `spack compiler list`.

**Comment by eschnett** [comment]
`list` vs. `find`: Spack has commands that work on packages (i.e. package descriptions), and commands that work on installs (i.e. stuff that is compiled etc.). Those are two different pair of shoes, and could use a different prefix to separate them:
- `spack pkg create`, `spack pkg list`, `spack pkg info` etc. for looking at package files
- `spack install`, `spack list`, `spack test` etc. for working with stuff that is installed

**Comment by healther** [comment]
Let me first comment on @adamjstewart's points:

- ```list``` vs ```find```: This is confusing as hell in the beginning and in theory I like @eschnett's idea of grouping similar commands together (more on that later), but since these two are maybe the most central elements of ```spack```, I don't think here it is appropriate. I personally like ```list``` and ```avail```, but I also knew modules before I knew spack... I still think, if the ```spack help``` command is somewhat clean, ```avail``` would be sufficiently far up, to be noticed by every user.

- ```load``` and co: It's nice that ```spack``` creates modules, so that one can expose ```module load <modulename>``` to users. However our usecase envisions (at least some) "users" to interact with ```spack``` directly. So it would be nice, to have a convenience function on the spack-level to load specs. I'll outline what my personal thoughts are below.

- ```clean``` vs ```purge```: Those two are basically both cleaning/resetting spack's internals -> I would like to have them grouped, and being able to be modified with flags. Regarding the cleaning of stages, I'd actually prefer that to be in the ```install``` command (see below)

- ```config``` & ```diy```: I'm not too sure what these are doing exactly.

- ```test``` & ```install --run-tests```: The trigger for spack's unit-test should probably not be a toplevel command at all.

- ```md5``` & ```checksum```: see below

- ```doc```: Not sure what this does

- ```aliases```: "There should be one-- and preferably only one --obvious way to do it."

Okay let's come at this from a different point of view: What does the user need to use spack? 

There are roughly 4 categories of spack functionality:
1. Repeatable installation and deployment of packages
2. Exposure of installed packages to users
3. Creation of new packages
4. Internals

I guess, but here I'm only guessing, that most users really only care about 1-3. 

1. Basically requires the functionality that is provided by ```install```, ```uninstall```, ```info```, ```spec```, ```list``` and ```find```. This is the core of spack and should be top-level commands (maybe not ```spec```, as it is a spack-unique abstraction)
2. I'd group here functionality like ```activate```, ```deactivate```, ```load```, ```unload```, ```module```, ```list```, ```find```, ```use``` and ```unuse```. This is no longer really the heart of spack, but it's still functionality that should be easily available. We would for example like something like ```spack load spec``` to be a valid use case. But maybe this all should be grouped into one subcommand.
3. ```create```, ```edit```, ```info```, ```checksum```, ```md5``` and friends. Should probably go into one subcommand like ```spack package create <packagename>```
4. Really all the rest, but especially the development tools (i.e. unit-tests, debug, flake8, python, etc) should go into one subcommand. Then there should be one to manage spack's state (configure, config, clean, purge, mirror, reindex)

For our use-case: Relatively small group, but no dedicated sys-admin personal, we would like to have a simple way of migrating to new systems (i.e. different cluster, new OS, etc). At the end there should/needs to be a set of software, defined at the moment in an extra ```package.py```, that must be able to be loaded by individual users. Preferably that system would be easily composable, think having multiple, possibly overlapping, sets of packages for different users.
Ideally users would retain the ability to add software on their own, but not influence existing installations.

So as a tl;dr:
I'm happy with the functionality and commands for 1.), Unhappy with 2.), would like to see 3.) moved to one subcommand and really confused by the amount of top-level commands from 4.)
For 2) I'd prefer to have functionality like ```spack load <spec1> <spec2> ... ``` just work, including to complain when asked to load non-existing or multiple or already loaded packages.

**Comment by citibeth** [comment]
Why does this discussion remind me of #2867, which was shut down within one day?  Maybe the efforts to improve `spack help` in #3033 have highlighted the issue?

**Comment by citibeth** [comment]
One of the problems with the current commands is they mix up what Spack *could* build with what it *has* built.  I believe that work on Spack Environments gives us an opportunity to better separate those two kinds of functionality --- because commands about what HAS been built will generally be with respect to a specific environment.

> list vs. find

These can be folded into one general concept of "Look in a bunch of objects, filtered based on the spec I give."  In one case, that objects being searched is a repo; in the other case, it's a bunch of installed packages (which will be represented by a "global" spec list that tracks everything that's been installed).

> Do we really need spack compiler list and spack compilers?

No.  Let's get rid of both and subsume into the `spack list` command, which will know how to list anything.    `spack list compilers` will list compilers, `spack list available-packages py-*` will list available packages starting with `py-`, and `spack --spec-list=mylist list installed` will list installed packages that are part of the `mylist` spec list. And so on...


> module load vs spack load vs spack module load

The problem with modules has always been finding the right one.  `spack load` does that; but using the slow/non-scriptable logic from `spack find`.  `spack module loads` allows you to cache that slow/non-scriptable logic and generate a bunch of `module load` commands instead.

All of this needs to be overhauled with Spack Environments.  Ideally, Spack will create a single super-module file that loads an entire environment, rather than one package at a time. That module will be named after the environment (which is a user-chosen name), and will not have hashes in its name.  With that change, `spack load` and `spack module load` will become less useful or deprecated.

> clean vs purge

It should be called `clean` because that's what Make does.  And anything else that cleans / garbage collects should be added to the command.  The syntax should be `spack clean <what-you-want-to-clean>`.  For example, what happens when you upgrade to new versions of packages and `zlib-1.0` is now no longer used by anything in your spec lists (environments)? A `spack clean unused-packages` should be able to collect that garbage.

> clean vs purge

"Build pipeline" options should be extended with a word.  Instead of `spack build`, it should be something like `spack stage=build`.

> diy vs setup

This will get revamped with Spack Environments; at that point, `spack setup` will probably no longer be needed as a separate command. Same for `diy`.  (Instead, you will specify in the spec list which packages you want to `setup` or `diy`).  @scheibelp has some neat ideas that will generalize this as well.

> test vs. install --run-tests

Let's demote `spack test` to a sub-command somewhere.  It's really a developer thing.

> doc vs docs

Looking at #3033 I don't know what `docs` does.  But I believe `spack doc` should be subsumed as a sub-command next to `spack test`.

> md5 vs checksum

yes, merge.  The command should be `spack checksum <thing-you-want-to-checksum>`, where the thing is a package, file, etc.  And you want different kinds of checksums, that should be an optional flag to `spack checksum`.

**Comment by tgamblin** [comment]
@citibeth: To be completely honest, #2867 helped motivate the help cleanup. But this issue is *far* more constructively worded.

**Comment by SteVwonder** [comment]
tl;dr: other really common package managers use `search` to query available packages and `list` to query installed packages.  Spack does the exact opposite and uses `list` to query available packages and `find` to query installed packages.  Great confusion ensues.

-----

Sorry to resurrect an old thread, but I continually stumble on `list` and `find`.

The hard part for me is the momentum from other package managers.  I expect the command that "searches" all available package to be `search`.  It is that way for `yum`, `apt`, `brew`, and others.  So I find myself always reaching for `spack search`, and when that fails, my brain reaches for `spack find` since I consider `find` to have a more similar meaning to `search` than `list` does.   To add to the frustration, when I run `spack find <package>` on a package that does exist but I don't have installed, the output is:
```
→ spack find flux-sched
==> No package matches the query: flux-sched
```
Which in no way tells me that the command searched my installed packages rather than the available packages.  So I'm left thinking Spack doesn't support the package at all, when really, I just don't have the package installed.  At this point, I find myself running `spack list <package>` if only to double check that I haven't run the wrong command (which I invariably have).

The same trouble applies when I want to "list" the packages that I have installed.  I reach for `spack list` since that is the correct command for `brew`, `yum`, and probably others.

I think it would greatly increase usability to use the same terminology as other package managers.  At minimum, aliasing `list` to `search` would eliminate 50% of my problems.  Additionally, adding text to the output of `list` and `find` that specifies whether spack queried `installed` or `available` packages would quickly alert me to the fact that I ran the wrong command.

**Comment by tgamblin** [comment]
> other really common package managers use search to query available packages and list to query installed packages

To be fair, it’s based on MacPorts and fink, which both use `list` to list available packages. Contrary to what you say above, `yum list` also lists available packages (though there is also `yum list installed`). I’m not really seeing where there’s any clear standard.

I think swapping them at this point would confuse a lot of people more than they already are.

But what if there was also `spack avail` or something clearer as a synonym for list?  Not sure if more options would really help, but I’d rather stay backward compatible on this one, people are sure to have scripted these.

**Comment by healther** [comment]
I would associate `avail` with `module avail` which is a list of _loadable_ modules, i.e. in spack-speak installed packages. 🙈 

The special problem for spack is that it depends on who you are. If you are a sys-admin `avail` means "what can spack eventually provide for me", if you are a user it means "what can I use right now". Intuitive naming requires a common reference frame, so while I would agree with what @adamjstewart said above. Achieving this kind of "well named" might be beyond the reach of spack.

But specifically for the confusion `find` and `list` we could just add a reference to the other command to the output. Something like
before:
```
→ spack find flux-sched
==> No package matches the query: flux-sched. 
```
after:
```
→ spack find flux-sched
==> No installed package matches the query: flux-sched. If you are looking for installable packages try `spack list flux-sched`
```

Yes being similar to different package managers makes adoption easier, but I'm not sure how large the advantage is over a really helpful UI.

**Comment by tgamblin** [comment]
@healther: want to do a PR for that? I’d put the second sentence on a new line (as the second arg to `tty.msg`)

**Comment by SteVwonder** [comment]
> To be fair, it’s based on MacPorts and fink, which both use list to list available packages.

Looking over MacPorts and Fink's commands, I think they are less confusing because they don't have a `find` command in addition to `list`.  They just add the `--installed` flag to `list`, which is very explicit and clear.  I actually really like that design.  It is hard to confuse commands when there is only one.  Maybe this functionality could be added via a `spack query` command (while retaining `find` and `list` for compatibility)?

> I’m not really seeing where there’s any clear standard.

One commonality that I'm seeing across `yum`, `dnf`, `port`, `apt`, and `brew` is that they all provide a `search` command that queries available packages (and possibly their descriptions).  It seems that Fink is the only one that doesn't provide `search` and instead opts for `fink apropos <string>`.

**Comment by SteVwonder** [comment]
> But specifically for the confusion find and list we could just add a reference to the other command to the output. Something like

That sounds good.  I also recommend modifying `spack list` to output the adjective `available` (or something else that distinguishes the output).   I would recommend `installable` here but that looks too similar to `installed` and could be easily mistaken.

before:
```
→ spack list asdf
==> 0 packages.
```

after:
```
→ spack list asdf
==> 0 available packages.
```

You could even use this as a opportunity to encourage contributing :)
```
→ spack list asdf
==> 0 available packages.  If you want to learn how to create a new package, visit https://spack.readthedocs.io/en/latest/contribution_guide.html
```

**Comment by healther** [comment]
@tgamblin @SteVwonder see #9165. I refrained from adding the "you can contribute" statement, as I'm not sure how much we should just blindly assume that any user should also be considered a potential contributor

**Comment by adamjstewart** [comment]
This discussion is pretty dead, so I'm going to close this.
